File: skills/skill-creator/scripts/run_eval.py
#!/usr/bin/env python3
"""Evaluate whether a compatible agent loads a skill for a set of queries."""

import argparse
import json
import os
import select
import shutil
import sys
import tempfile
import time
from concurrent.futures import ProcessPoolExecutor, as_completed
from pathlib import Path

from scripts.runners import RunnerError, create_runner
from scripts.utils import parse_skill_md


def find_project_root() -> Path:
    """Find the nearest project root using portable agent conventions."""
    current = Path.cwd()
    for parent in [current, *current.parents]:
        if (parent / ".agents").is_dir() or (parent / "AGENTS.md").is_file():
            return parent
        if (parent / ".git").exists():
            return parent
    return current


def _ignore_generated(_directory: str, names: list[str]) -> set[str]:
    return {name for name in names if name == "__pycache__" or name.endswith(".pyc")}


def _copy_skill(skill_path: Path, project_root: Path) -> tuple[Path, str]:
    """Copy a skill into an isolated project's canonical .agents/skills path."""
    name, _, _ = parse_skill_md(skill_path)
    target = project_root / ".agents" / "skills" / name
    target.parent.mkdir(parents=True, exist_ok=True)
    shutil.copytree(skill_path, target, ignore=_ignore_generated)
    return target, name


def run_single_query(
    query: str,
    skill_path: str,
    timeout: int,
    model: str | None = None,
    runner: str | None = None,
) -> bool:
    """Run one query in an isolated project and report whether the skill loaded."""
    source = Path(skill_path).resolve()
    with tempfile.TemporaryDirectory(prefix="skill-trigger-eval-") as tmp:
        project_root = Path(tmp)
        _, skill_name = _copy_skill(source, project_root)
        process = create_runner(runner).start_stream(query, model, project_root)
        if process.stdout is None:
            raise RunnerError("Runner did not expose stdout")

        start_time = time.time()
        buffer = ""
        try:
            while time.time() - start_time < timeout:
                if process.poll() is not None:
                    remaining = process.stdout.read()
                    if remaining:
                        buffer += remaining.decode("utf-8", errors="replace")
                    break

                ready, _, _ = select.select([process.stdout], [], [], 1.0)
                if not ready:
                    continue
                chunk = os.read(process.stdout.fileno(), 8192)
                if not chunk:
                    break
                buffer += chunk.decode("utf-8", errors="replace")

                while "\n" in buffer:
                    line, buffer = buffer.split("\n", 1)
                    try:
                        event = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    decision = create_runner(runner).event_loaded_skill(event, skill_name)
                    if decision is not None:
                        return decision

            if buffer.strip():
                try:
                    event = json.loads(buffer)
                    decision = create_runner(runner).event_loaded_skill(event, skill_name)
                    if decision is not None:
                        return decision
                except json.JSONDecodeError:
                    pass
            return False
        finally:
            if process.poll() is None:
                process.kill()
                process.wait()


def run_eval(
    eval_set: list[dict],
    skill_path: Path,
    num_workers: int,
    timeout: int,
    runs_per_query: int = 1,
    trigger_threshold: float = 0.5,
    model: str | None = None,
    runner: str | None = None,
    description: str | None = None,
) -> dict:
    """Run the full eval set and return aggregate results."""
    name, original_description, _ = parse_skill_md(skill_path)
    description = description or original_description
    results = []

    with tempfile.TemporaryDirectory(prefix="skill-description-") as tmp:
        staged_skill = Path(tmp) / skill_path.name
        shutil.copytree(skill_path, staged_skill, ignore=_ignore_generated)
        skill_md = staged_skill / "SKILL.md"
        content = skill_md.read_text()
        if original_description != description:
            lines = content.splitlines()
            frontmatter_end = next(
                (index for index, line in enumerate(lines[1:], start=1) if line.strip() == "---"),
                None,
            )
            if frontmatter_end is None:
                raise ValueError("Could not locate the SKILL.md frontmatter")
            description_index = next(
                (
                    index
                    for index, line in enumerate(lines[1:frontmatter_end], start=1)
                    if line.startswith("description:")
                ),
                None,
            )
            if description_index is None:
                raise ValueError("Could not locate the SKILL.md description for evaluation")
            description_end = description_index + 1
            if lines[description_index].split(":", 1)[1].strip() in (">", "|", ">-", "|-"):
                while description_end < frontmatter_end and lines[description_end].startswith(("  ", "\t")):
                    description_end += 1
            escaped_description = description.replace("'", "''")
            lines[description_index:description_end] = [f"description: '{escaped_description}'"]
            skill_md.write_text("\n".join(lines))

        with ProcessPoolExecutor(max_workers=num_workers) as executor:
            future_to_item = {}
            for item in eval_set:
                for _ in range(runs_per_query):
                    future = executor.submit(
                        run_single_query,
                        item["query"],
                        str(staged_skill),
                        timeout,
                        model,
                        runner,
                    )
                    future_to_item[future] = item

            query_triggers: dict[str, list[bool]] = {}
            query_items: dict[str, dict] = {}
            for future in as_completed(future_to_item):
                item = future_to_item[future]
                query = item["query"]
                query_items[query] = item
                query_triggers.setdefault(query, [])
                try:
                    query_triggers[query].append(future.result())
                except Exception as error:
                    print(f"Warning: query failed: {error}", file=sys.stderr)
                    query_triggers[query].append(False)

    for query, triggers in query_triggers.items():
        item = query_items[query]
        trigger_rate = sum(triggers) / len(triggers)
        should_trigger = item["should_trigger"]
        did_pass = trigger_rate >= trigger_threshold if should_trigger else trigger_rate < trigger_threshold
        results.append({
            "query": query,
            "should_trigger": should_trigger,
            "trigger_rate": trigger_rate,
            "triggers": sum(triggers),
            "runs": len(triggers),
            "pass": did_pass,
        })

    passed = sum(1 for result in results if result["pass"])
    return {
        "skill_name": name,
        "description": description,
        "results": results,
        "summary": {"total": len(results), "passed": passed, "failed": len(results) - passed},
    }


def main():
    parser = argparse.ArgumentParser(description="Run skill trigger evaluation")
    parser.add_argument("--eval-set", required=True, help="Path to eval set JSON file")
    parser.add_argument("--skill-path", required=True, help="Path to skill directory")
    parser.add_argument("--description", default=None, help="Override description to test")
    parser.add_argument("--num-workers", type=int, default=4, help="Number of parallel workers")
    parser.add_argument("--timeout", type=int, default=60, help="Timeout per query in seconds")
    parser.add_argument("--runs-per-query", type=int, default=3, help="Number of runs per query")
    parser.add_argument("--trigger-threshold", type=float, default=0.5, help="Trigger rate threshold")
    parser.add_argument("--model", default=None, help="Model override passed to the runner")
    parser.add_argument("--runner", choices=["goose"], default=None, help="Evaluation runner (default: SKILL_CREATOR_RUNNER or goose)")
    parser.add_argument("--verbose", action="store_true", help="Print progress to stderr")
    args = parser.parse_args()

    eval_set = json.loads(Path(args.eval_set).read_text())
    skill_path = Path(args.skill_path).resolve()
    if not (skill_path / "SKILL.md").exists():
        parser.error(f"No SKILL.md found at {skill_path}")

    try:
        output = run_eval(
            eval_set=eval_set,
            skill_path=skill_path,
            description=args.description,
            num_workers=args.num_workers,
            timeout=args.timeout,
            runs_per_query=args.runs_per_query,
            trigger_threshold=args.trigger_threshold,
            model=args.model,
            runner=args.runner,
        )
    except (RunnerError, ValueError) as error:
        parser.error(str(error))

    if args.verbose:
        summary = output["summary"]
        print(f"Results: {summary['passed']}/{summary['total']} passed", file=sys.stderr)
        for result in output["results"]:
            status = "PASS" if result["pass"] else "FAIL"
            print(
                f"  [{status}] rate={result['triggers']}/{result['runs']} "
                f"expected={result['should_trigger']}: {result['query'][:70]}",
                file=sys.stderr,
            )
    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()

File: skills/skill-creator/scripts/run_eval.ts
#!/usr/bin/env node
// Evaluate whether a compatible agent loads a skill for a set of queries.
import { mkdirSync, mkdtempSync, rmSync, readFileSync, writeFileSync, cpSync, existsSync } from "node:fs";
import { join, resolve, basename } from "node:path";
import { tmpdir } from "node:os";
import { parseArgs } from "node:util";
import { createRunner, RunnerError } from "./runners/index.js";
import { parseSkillMd } from "./utils.js";

interface EvalItem {
  query: string;
  should_trigger: boolean;
}

interface EvalResult {
  query: string;
  should_trigger: boolean;
  trigger_rate: number;
  triggers: number;
  runs: number;
  pass: boolean;
}

interface EvalOutput {
  skill_name: string;
  description: string;
  results: EvalResult[];
  summary: { total: number; passed: number; failed: number };
}

function copySkill(skillPath: string, projectRoot: string): { target: string; name: string } {
  const { name } = parseSkillMd(skillPath);
  const target = join(projectRoot, ".agents", "skills", name);
  mkdirSync(join(projectRoot, ".agents", "skills"), { recursive: true });
  cpSync(skillPath, target, {
    recursive: true,
    filter: (src) => !src.includes("__pycache__") && !src.endsWith(".pyc"),
  });
  return { target, name };
}

async function runSingleQuery(
  query: string,
  skillPath: string,
  timeoutMs: number,
  model: string | undefined,
  runner: string | undefined
): Promise<boolean> {
  const source = resolve(skillPath);
  const tmp = mkdtempSync(join(tmpdir(), "skill-trigger-eval-"));
  try {
    const { name: skillName } = copySkill(source, tmp);
    const runnerInstance = createRunner(runner);
    const child = runnerInstance.startStream(query, model ?? null, tmp);

    return await new Promise<boolean>((resolvePromise, reject) => {
      let buffer = "";
      let settled = false;
      const timer = setTimeout(() => {
        if (!settled) {
          settled = true;
          child.kill();
          resolvePromise(false);
        }
      }, timeoutMs);

      const processLine = (line: string) => {
        if (settled) return;
        try {
          const event = JSON.parse(line);
          const decision = runnerInstance.eventLoadedSkill(event, skillName);
          if (decision !== null) {
            settled = true;
            clearTimeout(timer);
            child.kill();
            resolvePromise(decision);
          }
        } catch {
          // not JSON, ignore
        }
      };

      child.stdout.on("data", (chunk: Buffer) => {
        buffer += chunk.toString("utf-8");
        let idx: number;
        while ((idx = buffer.indexOf("\n")) !== -1) {
          const line = buffer.slice(0, idx);
          buffer = buffer.slice(idx + 1);
          processLine(line);
        }
      });

      child.on("close", () => {
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          if (buffer.trim()) processLine(buffer);
          if (!settled) resolvePromise(false);
        }
      });

      child.on("error", (error) => {
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          reject(error);
        }
      });
    });
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }
}

async function mapLimit<T, R>(items: T[], limit: number, fn: (item: T) => Promise<R>): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let index = 0;
  async function worker() {
    while (index < items.length) {
      const current = index;
      index += 1;
      results[current] = await fn(items[current]);
    }
  }
  await Promise.all(Array.from({ length: Math.min(limit, items.length) }, worker));
  return results;
}

export async function runEval(
  evalSet: EvalItem[],
  skillPath: string,
  numWorkers: number,
  timeoutMs: number,
  options: {
    runsPerQuery?: number;
    triggerThreshold?: number;
    model?: string;
    runner?: string;
    description?: string;
  } = {}
): Promise<EvalOutput> {
  const {
    runsPerQuery = 1,
    triggerThreshold = 0.5,
    model,
    runner,
    description: descriptionOverride,
  } = options;

  const { name, description: originalDescription } = parseSkillMd(skillPath);
  const description = descriptionOverride || originalDescription;

  const tmp = mkdtempSync(join(tmpdir(), "skill-description-"));
  let queryTriggers: Map<string, boolean[]>;
  let queryItems: Map<string, EvalItem>;
  try {
    const stagedSkill = join(tmp, basename(skillPath));
    cpSync(skillPath, stagedSkill, {
      recursive: true,
      filter: (src) => !src.includes("__pycache__") && !src.endsWith(".pyc"),
    });
    const skillMdPath = join(stagedSkill, "SKILL.md");
    let content = readFileSync(skillMdPath, "utf-8");

    if (originalDescription !== description) {
      const lines = content.split("\n");
      let frontmatterEnd = -1;
      for (let i = 1; i < lines.length; i += 1) {
        if (lines[i].trim() === "---") {
          frontmatterEnd = i;
          break;
        }
      }
      if (frontmatterEnd === -1) {
        throw new Error("Could not locate the SKILL.md frontmatter");
      }
      let descriptionIndex = -1;
      for (let i = 1; i < frontmatterEnd; i += 1) {
        if (lines[i].startsWith("description:")) {
          descriptionIndex = i;
          break;
        }
      }
      if (descriptionIndex === -1) {
        throw new Error("Could not locate the SKILL.md description for evaluation");
      }
      let descriptionEnd = descriptionIndex + 1;
      const valueAfterColon = lines[descriptionIndex].split(":").slice(1).join(":").trim();
      if ([">", "|", ">-", "|-"].includes(valueAfterColon)) {
        while (
          descriptionEnd < frontmatterEnd &&
          (lines[descriptionEnd].startsWith("  ") || lines[descriptionEnd].startsWith("\t"))
        ) {
          descriptionEnd += 1;
        }
      }
      const escapedDescription = description.replace(/'/g, "''");
      lines.splice(descriptionIndex, descriptionEnd - descriptionIndex, `description: '${escapedDescription}'`);
      content = lines.join("\n");
      writeFileSync(skillMdPath, content, "utf-8");
    }

    queryTriggers = new Map();
    queryItems = new Map();

    const jobs: Array<{ item: EvalItem }> = [];
    for (const item of evalSet) {
      for (let i = 0; i < runsPerQuery; i += 1) {
        jobs.push({ item });
      }
    }

    await mapLimit(jobs, numWorkers, async ({ item }) => {
      queryItems.set(item.query, item);
      if (!queryTriggers.has(item.query)) queryTriggers.set(item.query, []);
      try {
        const result = await runSingleQuery(item.query, stagedSkill, timeoutMs, model, runner);
        queryTriggers.get(item.query)!.push(result);
      } catch (error) {
        console.error(`Warning: query failed: ${(error as Error).message}`);
        queryTriggers.get(item.query)!.push(false);
      }
    });
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }

  const results: EvalResult[] = [];
  for (const [query, triggers] of queryTriggers.entries()) {
    const item = queryItems.get(query)!;
    const triggerRate = triggers.filter(Boolean).length / triggers.length;
    const shouldTrigger = item.should_trigger;
    const didPass = shouldTrigger ? triggerRate >= triggerThreshold : triggerRate < triggerThreshold;
    results.push({
      query,
      should_trigger: shouldTrigger,
      trigger_rate: triggerRate,
      triggers: triggers.filter(Boolean).length,
      runs: triggers.length,
      pass: didPass,
    });
  }

  const passed = results.filter((r) => r.pass).length;
  return {
    skill_name: name,
    description,
    results,
    summary: { total: results.length, passed, failed: results.length - passed },
  };
}

async function main() {
  const { values } = parseArgs({
    args: process.argv.slice(2),
    options: {
      "eval-set": { type: "string" },
      "skill-path": { type: "string" },
      description: { type: "string" },
      "num-workers": { type: "string", default: "4" },
      timeout: { type: "string", default: "60" },
      "runs-per-query": { type: "string", default: "3" },
      "trigger-threshold": { type: "string", default: "0.5" },
      model: { type: "string" },
      runner: { type: "string" },
      verbose: { type: "boolean", default: false },
    },
  });

  if (!values["eval-set"] || !values["skill-path"]) {
    console.error("usage: run_eval.js --eval-set <file> --skill-path <dir> [options]");
    process.exit(2);
  }

  const evalSet: EvalItem[] = JSON.parse(readFileSync(resolve(values["eval-set"] as string), "utf-8"));
  const skillPath = resolve(values["skill-path"] as string);
  if (!existsSync(join(skillPath, "SKILL.md"))) {
    console.error(`error: No SKILL.md found at ${skillPath}`);
    process.exit(2);
  }

  let output: EvalOutput;
  try {
    output = await runEval(
      evalSet,
      skillPath,
      Number(values["num-workers"]),
      Number(values.timeout) * 1000,
      {
        description: values.description as string | undefined,
        runsPerQuery: Number(values["runs-per-query"]),
        triggerThreshold: Number(values["trigger-threshold"]),
        model: values.model as string | undefined,
        runner: values.runner as string | undefined,
      }
    );
  } catch (error) {
    if (error instanceof RunnerError || error instanceof Error) {
      console.error(`error: ${error.message}`);
      process.exit(2);
    }
    throw error;
  }

  if (values.verbose) {
    console.error(`Results: ${output.summary.passed}/${output.summary.total} passed`);
    for (const result of output.results) {
      const status = result.pass ? "PASS" : "FAIL";
      console.error(
        `  [${status}] rate=${result.triggers}/${result.runs} expected=${result.should_trigger}: ${result.query.slice(0, 70)}`
      );
    }
  }
  console.log(JSON.stringify(output, null, 2));
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error(error?.message ?? error);
    process.exit(1);
  });
}
